libs/s25main/mapGenerator/Resources.cpp

```diff
@@ -177,39 +177,65 @@ namespace rttr { namespace mapGenerator {
     {
         const auto& textures = map.textureMap;
         auto& resources = map.resources;
+        struct mineableResourceInfo
+        {
+            // The current "budget", that is, how many many are we over, or under, compared to the desired distribution.
+            int budget = 0;
+            // The ratio is the user defined ratio for this resource.
+            int ratio;
+            // The resource encoding
+            libsiedler2::Resource resource;
+            mineableResourceInfo(int ratio, libsiedler2::Resource resource) : ratio(ratio), resource(resource) {}
+        };
+        mineableResourceInfo mRIs[4] = {mineableResourceInfo(settings.ratioCoal, libsiedler2::R_Coal),
+                                        mineableResourceInfo(settings.ratioGold, libsiedler2::R_Gold),
+                                        mineableResourceInfo(settings.ratioIron, libsiedler2::R_Iron),
+                                        mineableResourceInfo(settings.ratioGranite, libsiedler2::R_Granite)};
         int total = settings.ratioCoal + settings.ratioGold + settings.ratioIron + settings.ratioGranite;
 
         RTTR_FOREACH_PT(MapPoint, map.size)
         {
             if(textures.All(pt, IsMinableMountain))
             {
+                int ratio = 0;
                 int randomNumber = rnd.RandomValue(1, total);
-                int ratio = settings.ratioGold;
-
-                if(randomNumber < ratio)
-                {
-                    resources[pt] = libsiedler2::R_Gold + rnd.RandomValue(0, 8);
-                    continue;
-                }
-
-                ratio += settings.ratioCoal;
-                if(randomNumber < ratio)
-                {
-                    resources[pt] = libsiedler2::R_Coal + rnd.RandomValue(0, 8);
-                    continue;
-                }
-
-                ratio += settings.ratioIron;
-                if(randomNumber < ratio)
-                {
-                    resources[pt] = libsiedler2::R_Iron + rnd.RandomValue(0, 8);
-                    continue;
-                }
-
-                ratio += settings.ratioGranite;
-                if(randomNumber < ratio)
+                for(auto mRI : mRIs)
                 {
-                    resources[pt] = libsiedler2::R_Granite + rnd.RandomValue(0, 8);
+                    ratio += mRI.ratio;
+
+                    // Wrong resource, look further.
+                    if(randomNumber > ratio)
+                        continue;
+
+                    // Right resource, adjust and check the budget.
+                    ++mRI.budget;
+                    if(mRI.budget <= 0)
+                        break;
```

Comment:
This line not being covered looks like there is a bug in the algorithm, I'd expect this to happen quite often, but then again: If you got the "right resource" but it is over budget, isn't it possible, that NO resource will be placed here?

Reply:
yes it is. and that is how -4 later comes into play. The original also had empty fields, which I don't find to be problematic.

Reply:
the not being covered part might just be the test?

Reply:
Just thinking: you pick Randomly the most common res. Place 16 fields of great below, then the next time this resource is picked it should be over budget. But according to coverage it is not. Our am I missing anything?

Reply:
what coverage? 
A resource is randomly picked, placed in a cluster and for each placement budget is decremented by one. So in the end it might be -5. The next 5 times the resource is picked we will not place it but just even out the budget. After that, a new cluster is placed for the resource.

Reply:
The break here should be hit if the budget is ever negative. But there was a note here that this line is not covered (check the files view once ci ran through) so it looks like this situation which should occur does not. 
Does that explain what I mean? 

Maybe check in the debugger to make sure this is hit or check back here after ci passed

Reply:
Last version was missing a reference, now it hits the continue if you create a random map (so budget goes below 0).
![2021-03-01_16-03-1614636192](https://user-images.githubusercontent.com/6248191/109565134-a3226700-7aa7-11eb-9b2d-e997c09efe31.jpg)